apps/rest_api/lib/controllers/public/auth.ex

```diff
@@ -64,19 +66,41 @@ defmodule RestAPI.Controllers.Public.Auth do
     params = Map.merge(params, conn.private.tracking)
 
     with {:ok, input} <- AuthorizationCode.cast_and_apply(params),
-         {:ok, response} <- Commands.sign_in_authorization_code(input) do
+         {:ok, response} <- Authenticator.sign_in_authorization_code(input) do
       conn
       |> put_view(Auth)
       |> put_status(200)
       |> render("token.json", response: response)
     end
   end
 
+  @doc """
+  Authorize an client to sign in the user later by using an authorization code.
+
+  This flow should only be used on authorization code grants.
+  """
+  @spec authorize(conn :: Plug.Conn.t(), params :: map()) :: Plug.Conn.t()
+  def authorize(%{private: %{session: session}} = conn, params) do
+    with {:ok, input} <- AuthorizationCodeSignIn.cast_and_apply(params),
+         {:ok, resp} <- Authorizer.authorize_authorization_code_sign_in(input, session.subject_id) do
+      if input.redirect_uri do
+        redirect(conn,
+          external: "#{input.redirect_uri}?code=#{resp.authorization_code}&state=#{input.state}"
+        )
+      else
+        conn
+        |> put_view(Auth)
+        |> put_status(200)
+        |> render("authorize.json", response: resp, state: input.state)
+      end
+    end
+  end
```

Comment:
```suggestion
    def authorize(%{private: %{session: session}} = conn, params) do
      with {:ok, %{redirect_uri: redirect_uri, state: state} = input} <-
             AuthorizationCodeSignIn.cast_and_apply(params),
           {:ok, resp} <-
             Authorizer.authorize_authorization_code_sign_in(input, session.subject_id) do
        do_authorize(redirect_uri, state, resp)
      end
    end

    defp do_authorize(nil, state, resp) do
      conn
      |> put_view(Auth)
      |> put_status(200)
      |> render("authorize.json", response: resp, state: state)
    end

    defp do_authorize(redirect_uri, state, %{authorization_code: authorization_code}) do
      redirect(conn,
        external: "#{redirect_uri}?code=#{resp.authorization_code}&state=#{state}"
      )
    end
  end
```

Reply:
I really don't think it's necessary to break every struct in matches and extract the value into variables only for future use when it's simple as it is.